src/raft_log.rs

```diff
@@ -495,14 +532,23 @@ impl<T: Storage> RaftLog<T> {
             snapshot_index = snapshot.get_metadata().index,
             snapshot_term = snapshot.get_metadata().term,
         );
-        self.committed = snapshot.get_metadata().index;
+        let index = snapshot.get_metadata().index;
+        assert!(index >= self.committed, "{} < {}", index, self.committed);
+        self.committed = index;
+        // persisted must be less than unstable.offset at any time.
+        if self.persisted > index {
+            self.persisted = index;
```

Comment:
The invariant of `persisted` is `persisted < unstable.offset && applied <= persisted`. So here it should be `index`.(The `unstable.offset` is `index + 1` and the `applied` will be `index` after applying this snapshot.
By the way, I think it twice and remove this `if` statement.

Reply:
I suggested the changes based on its meanings instead of invariant. It's should indicate that any logs before the index have been persisted. Snapshot at `index` is apparently not the case.

Reply:
I think about it before. Maybe we should have a better name?
We need `persisted` because the entries need to be persisted first and then they can be applied.
But for snapshot, the behavior of persisting and applying is only one step. That's the difference between them.